# Data Wrangling

<h5>

**Wrangling** /ˈræŋ.ɡəl.ɪŋ/

the activity of taking care of, controlling, or moving animals, especially large animals such as cows or horses

([Cambridge Dictionary](https://dictionary.cambridge.org/dictionary/english/wrangling))

</h5>

![Cattle Wrangler - image from https://commons.wikimedia.org/wiki/File:Pioneer_Day_Wrangler.jpg](https://upload.wikimedia.org/wikipedia/commons/thumb/8/83/Pioneer_Day_Wrangler.jpg/320px-Pioneer_Day_Wrangler.jpg)

**[Data wrangling](https://en.wikipedia.org/wiki/Data_wrangling)** commonly refers to the transformation of data from one "input" format (e.g., `.csv` files from an experiment), to a different format (e.g., a tidy dataframe) that is more appropriate to the needs of an analysis. In the context of the ExPra experiments, you will use data wrangling techniques to implement the transformations and data cleaning steps specified in your preregistrations.

## Setup

### Setup Part 1: Install Packages

We will use [`tidyverse`](https://www.tidyverse.org/) packages to implement our data wrangling. The "tidyverse" is a series of packages which share a philosphy based around code and data structures that are (a) tidy, and (b) readable. You can install the tidyverse packages like so:

```
install.packages("tidyverse")
```

This includes many packages that we won't be using today, but which will be useful in other parts of the course (e.g., on Data Visualisation).

Remember, you should install packages in the console - never in a script that your share with others. This is because otherwise, your script will go to the effort of reinstalling a package *every time* it is run!

### Setup Part 2: Check the Packages Load

Now, we can test that the packages we will be using today actually load. You should be able to run this code without any errors:

In [2]:
options(repr.plot.width=3.5, repr.plot.height=3)

In [3]:
library(dplyr)
library(tidyr)

### Setup Part 3

Finally, check that you can access the dataset we'll be using in this session. The `starwars` dataset is a dataset built into R that contains details of characters from the Star Wars films:

In [4]:
print(starwars)

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Luke Sk~    172    77 blond      fair       blue            19   male  mascu~
 2 C-3PO       167    75 NA         gold       yellow         112   none  mascu~
 3 R2-D2        96    32 NA         white, bl~ red             33   none  mascu~
 4 Darth V~    202   136 none       white      yellow          41.9 male  mascu~
 5 Leia Or~    150    49 brown      light      brown           19   fema~ femin~
 6 Owen La~    178   120 brown, gr~ light      blue            52   male  mascu~
 7 Beru Wh~    165    75 brown      light      blue            47   fema~ femin~
 8 R5-D4        97    32 NA         white, red red             NA   none  mascu~
 9 Biggs D~    183    84 black      light      brown           24   male  mascu~
10 Obi-Wan~    182    77 auburn, w~ fair       blue-gray       57   male  mascu~
# i 77 m

This is snapshot shows an example of tidy data - a philosophy of organising data such that each observation (*character*) has a single row, with all variables tied to that character as a single column.

Now that we're all set up, let's dig into the data...

## Arranging

We can use the `arrange()` function to sort by variables in the dataframe. For example, we can arrange all characters in order of height (shortest to tallest) like so:

In [5]:
arrange(starwars, height)

name                  height mass hair_color skin_color         
1   Yoda                   66    17.0 white      green              
2   Ratts Tyerell          79    15.0 none       grey, blue         
3   Wicket Systri Warrick  88    20.0 brown      brown              
4   Dud Bolt               94    45.0 none       blue, grey         
5   R2-D2                  96    32.0 NA         white, blue        
6   R4-P17                 96      NA none       silver, red        
7   R5-D4                  97    32.0 NA         white, red         
8   Sebulba               112    40.0 none       grey, red          
9   Gasgano               122      NA none       white, blue        
10  Watto                 137      NA black      blue, grey         
11  Leia Organa           150    49.0 brown      light              
12  Mon Mothma            150      NA auburn     fair               
13  Cordé                 157      NA brown      light              
14  Nien Nunb             160    68.0 none       grey               
15  Shmi Skywalker        163      NA black      fair               
16  Ben Quadinaros        163    65.0 none       grey, green, yellow
17  Beru Whitesun lars    165    75.0 brown      light              
18  Dormé                 165      NA brown      light              
19  Padmé Amidala         165    45.0 brown      light              
20  Barriss Offee         166    50.0 black      yellow             
21  C-3PO                 167    75.0 NA         gold               
22  Jocasta Nu            167      NA white      fair               
23  Zam Wesell            168    55.0 blonde     fair, green, yellow
24  Wedge Antilles        170    77.0 brown      fair               
25  Palpatine             170    75.0 grey       pale               
26  Finis Valorum         170      NA blond      fair               
27  Luminara Unduli       170    56.2 black      yellow             
28  Eeth Koth             171      NA black      brown              
29  Luke Skywalker        172    77.0 blond      fair               
30  Greedo                173    74.0 NA         green              
... ...                   ...    ...  ...        ...                
58  Raymus Antilles       188     79  brown      light              
59  Bossk                 190    113  none       green              
60  Nute Gunray           191     90  none       mottled green      
61  Bail Prestor Organa   191     NA  black      tan                
62  San Hill              191     NA  none       grey               
63  Qui-Gon Jinn          193     89  brown      fair               
64  Dooku                 193     80  white      fair               
65  Wat Tambor            193     48  none       green, grey        
66  Jar Jar Binks         196     66  none       orange             
67  Kit Fisto             196     87  none       green              
68  Mas Amedda            196     NA  none       blue               
69  Ki-Adi-Mundi          198     82  white      pale               
70  Dexter Jettster       198    102  none       brown              
71  IG-88                 200    140  none       metal              
72  Darth Vader           202    136  none       white              
73  Rugor Nass            206     NA  none       green              
74  Tion Medon            206     80  none       grey               
75  Taun We               213     NA  none       grey               
76  Grievous              216    159  none       brown, white       
77  Roos Tarpals          224     82  none       grey               
78  Chewbacca             228    112  brown      unknown            
79  Lama Su               229     88  none       grey               
80  Tarfful               234    136  brown      brown              
81  Yarael Poof           264     NA  none       white              
82  Arvel Crynyd           NA     NA  brown      fair               
83  Finn                   NA     NA  black      dark               
84

In [6]:
arrange(starwars, height) |> print()

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Yoda         66    17 white      green      brown            896 male  mascu~
 2 Ratts T~     79    15 none       grey, blue unknown           NA male  mascu~
 3 Wicket ~     88    20 brown      brown      brown              8 male  mascu~
 4 Dud Bolt     94    45 none       blue, grey yellow            NA male  mascu~
 5 R2-D2        96    32 NA         white, bl~ red               33 none  mascu~
 6 R4-P17       96    NA none       silver, r~ red, blue         NA none  femin~
 7 R5-D4        97    32 NA         white, red red               NA none  mascu~
 8 Sebulba     112    40 none       grey, red  orange            NA male  mascu~
 9 Gasgano     122    NA none       white, bl~ black             NA male  mascu~
10 Watto       137    NA black      blue, grey yellow            NA male  mascu~
# i 77 m

The `arrange()` function, like most `dplyr` verb functions, takes the dataframe (`starwars`) as its first argument, and the variable names (e.g., `height`) as subsequent arguments.

In height order, we can see that Yoda is the shortest character, at 66 cm, with podracer [Ratts Tyerell](https://starwars.fandom.com/wiki/Ratts_Tyerell) next shortest, at 79 cm.

We can also arrange the dataframe in *descending* order, with the *`desc()`* function.

In [7]:
arrange(starwars, desc(height))

name                  height mass hair_color skin_color         
1   Yarael Poof           264     NA  none       white              
2   Tarfful               234    136  brown      brown              
3   Lama Su               229     88  none       grey               
4   Chewbacca             228    112  brown      unknown            
5   Roos Tarpals          224     82  none       grey               
6   Grievous              216    159  none       brown, white       
7   Taun We               213     NA  none       grey               
8   Rugor Nass            206     NA  none       green              
9   Tion Medon            206     80  none       grey               
10  Darth Vader           202    136  none       white              
11  IG-88                 200    140  none       metal              
12  Ki-Adi-Mundi          198     82  white      pale               
13  Dexter Jettster       198    102  none       brown              
14  Jar Jar Binks         196     66  none       orange             
15  Kit Fisto             196     87  none       green              
16  Mas Amedda            196     NA  none       blue               
17  Qui-Gon Jinn          193     89  brown      fair               
18  Dooku                 193     80  white      fair               
19  Wat Tambor            193     48  none       green, grey        
20  Nute Gunray           191     90  none       mottled green      
21  Bail Prestor Organa   191     NA  black      tan                
22  San Hill              191     NA  none       grey               
23  Bossk                 190    113  none       green              
24  Anakin Skywalker      188     84  blond      fair               
25  Mace Windu            188     84  none       dark               
26  Saesee Tiin           188     NA  none       pale               
27  Plo Koon              188     80  none       orange             
28  Raymus Antilles       188     79  brown      light              
29  Gregar Typho          185     85  black      dark               
30  Adi Gallia            184     50  none       dark               
... ...                   ...    ...  ...        ...                
58  Luminara Unduli       170    56.2 black      yellow             
59  Zam Wesell            168    55.0 blonde     fair, green, yellow
60  C-3PO                 167    75.0 NA         gold               
61  Jocasta Nu            167      NA white      fair               
62  Barriss Offee         166    50.0 black      yellow             
63  Beru Whitesun lars    165    75.0 brown      light              
64  Dormé                 165      NA brown      light              
65  Padmé Amidala         165    45.0 brown      light              
66  Shmi Skywalker        163      NA black      fair               
67  Ben Quadinaros        163    65.0 none       grey, green, yellow
68  Nien Nunb             160    68.0 none       grey               
69  Cordé                 157      NA brown      light              
70  Leia Organa           150    49.0 brown      light              
71  Mon Mothma            150      NA auburn     fair               
72  Watto                 137      NA black      blue, grey         
73  Gasgano               122      NA none       white, blue        
74  Sebulba               112    40.0 none       grey, red          
75  R5-D4                  97    32.0 NA         white, red         
76  R2-D2                  96    32.0 NA         white, blue        
77  R4-P17                 96      NA none       silver, red        
78  Dud Bolt               94    45.0 none       blue, grey         
79  Wicket Systri Warrick  88    20.0 brown      brown              
80  Ratts Tyerell          79    15.0 none       grey, blue         
81  Yoda                   66    17.0 white      green              
82  Arvel Crynyd           NA      NA brown      fair               
83  Finn                   NA      NA black      dark               
84

In [11]:
arrange(starwars, desc(height)) |> print()

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Yarael ~    264    NA none       white      yellow          NA   male  mascu~
 2 Tarfful     234   136 brown      brown      blue            NA   male  mascu~
 3 Lama Su     229    88 none       grey       black           NA   male  mascu~
 4 Chewbac~    228   112 brown      unknown    blue           200   male  mascu~
 5 Roos Ta~    224    82 none       grey       orange          NA   male  mascu~
 6 Grievous    216   159 none       brown, wh~ green, y~       NA   male  mascu~
 7 Taun We     213    NA none       grey       black           NA   fema~ femin~
 8 Rugor N~    206    NA none       green      orange          NA   male  mascu~
 9 Tion Me~    206    80 none       grey       black           NA   male  mascu~
10 Darth V~    202   136 none       white      yellow          41.9 male  mascu~
# i 77 m

This shows us that long-necked Jedi, [Yarael Poof](https://starwars.fandom.com/wiki/Yarael_Poof), is the tallest character, at 264 cm.

As well as sorting by numeric variables, we can also sort by character variables. For instance, we can sort by hair colour. If we do this we will sort hair alphabetically, from "auburn" to "white".

In [ ]:
arrange(starwars, hair_color)

In [ ]:
arrange(starwars, hair_color) |> print()

Finally, we can sort by multiple variables at once. For instance, we may want to sort by hair colour alphabetically, and then by height descendingly:

In [ ]:
arrange(starwars, hair_color, desc(height))

In [ ]:
arrange(starwars, hair_color, desc(height)) |> print()